exporters/trace/ocagent/src/main/java/io/opencensus/exporter/trace/ocagent/OcAgentTraceServiceStreamObservers.java

```diff
@@ -0,0 +1,245 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/*>>>
+import org.checkerframework.checker.nullness.qual.Nullable;
+*/
+
+/** Stream observers of OC-Agent Trace Service. */
+final class OcAgentTraceServiceStreamObservers {
+
+  private OcAgentTraceServiceStreamObservers() {}
+
+  private static final Logger logger =
+      Logger.getLogger(OcAgentTraceServiceStreamObservers.class.getName());
+
+  /*
+   * We should use ManagedChannel.getState() to determine the state of current connection.
+   *
+   * // TODO(songya): avoid going into too much details of each stream and use generic
+   * request/response if that's more clear.
+   *
+   * 0. For export RPC we hold a reference "exportRequestObserverRef" to a StreamObserver for
+   * request stream, and "exportResponseObserverRef" to a StreamObserver for response stream.
+   * For config RPC we have "currentConfigObserverRef" and "updatedConfigObserverRef". Note that
+   * exportRequestObserver and currentConfigObserver are defined by Agent, while
+   * exportResponseObserver and updatedConfigObserver are defined in this class.
+   *
+   * Think of the whole process as a Finite State Machine.
+   *
+   * Note: "Exporter" = Agent Exporter and "Agent" = OC-Agent in the sections below.
+   *
+   * STATE 1: Unconnected/Disconnected.
+   *
+   * 1. First, Exporter will try to establish the streams by calling TraceServiceStub.export and
+   * TraceServiceStub.config in a daemon thread.
+   *
+   *   1-1. If the initial attempt succeeded, references of exportRequestObserver and
+   *   currentConfigObserver should be returned to Exporter as callbacks. We keep a reference of
+   *   them in exportRequestObserverRef and currentConfigObserverRef. At the same time, references
+   *   of exportRequestObserver and updatedConfigObserver will be passed to Agent as callbacks.
+   *
+   *   1-2. If the attempt failed, TraceServiceStub.export or TraceServiceStub.config will throw
+   *   an exception. We should catch the exceptions and keep retrying. References of
+   *   exportRequestObserver and currentConfigObserver must be null.
+   *
+   *   1-3. After each attempt, we should check if the connection succeeded or not with
+   *   ManagedChannel.getState(). (This is done in the daemon thread.)
+   *   If connection succeeded and Exporter has the references to exportRequestObserver and
+   *   currentConfigObserver, we can move forward and start sending/receiving streams
+   *   (move to STATE 2). Otherwise Exporter should retry.
+   *
+   * STATE 2: Already connected.
+   *
+   * 2. Once streams are open, they should be kept alive. Exporter sends spans to Agent by calling
+   * exportRequestObserver.onNext (the callback provided by Agent). In addition, in another daemon
+   * thread Exporter watches config updates from Agent, which is done by Agent calling back
+   * updatedConfigObserver.onNext.
+   *
+   *   2-1. If for some reason the connection is interrupted, the errors will be caught by
```

Comment:
Unify 2-1 and 2-2 by saying this:
If for some reason the connection is interrupted, the errors will be caught by the onError method of any stream observer.  In that case, we should reset references to all the stream observers.

Reply:
Done - also onComplete and onError are a bit different so there won't always be errors when the streams are disconnected.